director/projects/api/views/nodes.py

```diff
@@ -0,0 +1,222 @@
+        except Node.DoesNotExist:
+            node = Node.objects.create(
+                key=key, project=project, app=app, host=host, json=node
+            )
+
+        serializer = NodesCreateResponse(node, context={"request": request})
+        return Response(serializer.data, status=status.HTTP_201_CREATED,)
+
+    @swagger_auto_schema(responses={status.HTTP_200_OK: NodeSerializer},)
+    def retrieve(
+        self, request: Request, key: str, format: Optional[str] = None
+    ) -> Response:
+        """
+        Retrieve a node.
+
+        Performs content negotiation and authorization based on the `Accept` header.
+
+        - For `application/json` returns 403 if the request user is not authorized
+        to read the project.
+        - For `text/html` (and others) returns a very simple HTML rendering of the
+        of the node (only type etc) if the user does not have read access to the project.
+        Otherwise, should return a full HTML rendering of the node using Encoda.
+        """
+        node = get_object_or_404(Node, key=key)
+        accept = request.META.get("HTTP_ACCEPT")
+        if format == "json" or (accept and "application/json" in accept):
+            # Require the user is authenticated and has VIEW permissions for the project
+            if not request.user.is_authenticated:
+                raise NotAuthenticated
+            if not self.is_permitted(
+                request.user, ProjectPermissionType.VIEW, pk=node.project.id
+            ):
+                raise PermissionDenied
+
+            serializer = NodeSerializer(node, context={"request": request})
+            return Response(serializer.data)
+        else:
+            if self.is_permitted(
+                request.user, ProjectPermissionType.VIEW, pk=node.project.id
+            ):
+                # Return a more complete view if the user has VIEW permissions.
+                # This should include public projects.
+                try:
+                    # Currently allow this to fail if the converter binary
+                    # can not be found e.g. during CI testing
+                    conversion = ConverterFacade().convert(
+                        input_data=ConverterIo(
+                            ConverterIoType.PIPE,
+                            json.dumps(node.json).encode("utf8"),
+                            ConversionFormatId.json,
+                        ),
+                        output_data=ConverterIo(
+                            ConverterIoType.PIPE, None, ConversionFormatId.html
+                        ),
+                        context=ConverterContext(standalone=False),
+                    )
+                    html = conversion.stdout.decode("utf8")
+                except FileNotFoundError:
+                    html = ""
+
+                app_url, app_name = None, None
+                if node.app == "stencila:encoda":
+                    app_name = "Stencila Encoda"
```

Comment:
Can you make `app_name` and `app_url` values constants at the top of the file.

Reply:
Yep, done